File: bumperbot_ws/src/bumperbot_cpp_examples/src/simple_turtlesim_kinematics.cpp
#include "bumperbot_cpp_examples/simple_turtlesim_kinematics.hpp"

using std::placeholders::_1;

SimpleTurtlesimKinematics::SimpleTurtlesimKinematics(const std::string &name): Node(name) {
  turtle1_pose_sub = create_subscription<turtlesim::msg::Pose>("/turtle1/pose", 10, std::bind(&SimpleTurtlesimKinematics::turtle1PoseCallback, this, _1));
  turtle2_pose_sub = create_subscription<turtlesim::msg::Pose>("/turtle2/pose", 10, std::bind(&SimpleTurtlesimKinematics::turtle2PoseCallback, this, _1));
}

void SimpleTurtlesimKinematics::turtle1PoseCallback(const turtlesim::msg::Pose &pose) {
  last_turtle1_pose = pose;
}

void SimpleTurtlesimKinematics::turtle2PoseCallback(const turtlesim::msg::Pose &pose) {
  last_turtle2_pose = pose;

  float Tx = last_turtle2_pose.x - last_turtle1_pose.x;
  float Ty = last_turtle2_pose.y - last_turtle1_pose.y;

  const float pi = 3.141592f;

  float theta_rad = last_turtle2_pose.theta - last_turtle1_pose.theta;
  float theta_deg = 180.0f * theta_rad / pi;

  RCLCPP_INFO_STREAM(get_logger(), "vector Turtle1->Turtle2: (" << Tx << "," << Ty << ")\n" 
  << "[" << std::cos(theta_rad) << ",\t" << -std::sin(theta_rad) << "]\n" 
  << "["<< std::sin(theta_rad) << ",\t" << std::cos(theta_rad) << "]\n" );
}

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleTurtlesimKinematics>("simple_turtlesim_kinematics");
  rclcpp::spin(node);
  rclcpp::shutdown();
}

File: bumperbot_ws/src/bumperbot_py_examples/bumperbot_py_examples/simple_turtlesim_kinematics.py
import rclpy
from rclpy.node import Node
from turtlesim.msg import Pose
from math import pi, cos, sin

class SimpleTurtleSimKinematics(Node):
  def __init__(self):
    super().__init__("simple_turtlesim_kinematics")
    self.t1_pose_sub = self.create_subscription(
      msg_type=Pose, 
      topic="/turtle1/pose", 
      callback=self.turtle1PoseCallback,
      qos_profile=10)
    
    self.t2_pose_sub = self.create_subscription(
      msg_type=Pose, 
      topic="/turtle2/pose", 
      callback=self.turtle2PoseCallback,
      qos_profile=10)
    
    self.last_turtle1_pose = Pose()
    self.last_turtle2_pose = Pose()

  def turtle1PoseCallback(self, msg: Pose):
    self.last_turtle1_pose = msg

  def turtle2PoseCallback(self, msg: Pose):
    self.last_turtle2_pose = msg

    Tx = self.last_turtle2_pose.x - self.last_turtle1_pose.x
    Ty = self.last_turtle2_pose.y - self.last_turtle1_pose.y

    theta_rad = self.last_turtle2_pose.theta - self.last_turtle1_pose.theta
    theta_deg = 180.0 * theta_rad / pi

    self.get_logger().info(f"Vector[t1 -> t2]:\n ({Tx:.3}, {Ty:.3}, {theta_deg}', \n{cos(theta_rad)} {-sin(theta_rad)}\n {sin(theta_rad)} {cos(theta_rad)})\n")


def main():
  rclpy.init()
  node = SimpleTurtleSimKinematics()
  rclpy.spin(node)
  node.destroy_node()
  rclpy.shutdown()


if __name__ == "__main__":
  main()